src/main/java/com/squareup/kotlinpoet/TypeName.kt

```diff
@@ -64,9 +64,12 @@ import kotlin.reflect.KVariance
  * Create new reference types like `com.example.HelloWorld` with [ClassName.bestGuess]. To build composite
  * types like `Set<Long>`, use the factory methods on [ParameterizedTypeName], [TypeVariableName],
  * and [WildcardTypeName].
+ *
+ * @param S the self-type, allowing [copy] to return the subclass type
  */
-sealed class TypeName constructor(
-  val nullable: Boolean, annotations: List<AnnotationSpec>
+sealed class TypeName<out S : TypeName<S>> constructor(
```

Comment:
Ick. I don’t think the cost is worth the benefit. Would rather we just override all the corresponding methods to narrow the return type

Reply:
Agree, it's super noisy.. Will work on v2